text/1025-branches.rst

```diff
@@ -0,0 +1,181 @@
+   having database states that correspond to the schema and avoiding
+   accidental mismatches in migrations.
+
+   If the ``from <oldbranch>`` part is omitted, the new branch is created
+   using the branch from the current connection.
+
+3) ``drop branch <oldbranch>``
+
+   Removes an existing branch from the instance.
+
+
+Project and CLI Integration
+---------------------------
+
+Currently the project configuration file ``edgedb.toml`` does not have to
+specify the default *database* it connects to. With this proposal specifying
+the branch is mandatory.
+
+A new project should ask the user for the branch name defaulting to "main".
+
+The ``database`` CLI commands should be deprecated and new ``branch`` commands
+introduced instead.
+
+``edgedb branch create`` command options:
+
+* It must have the ``--from <oldname>`` option. If this option is omitted the
+  current branch specified in the ``edgedb.toml`` file is used as the "from"
+  branch. After the branch is created, the ``edgedb.toml`` file is updated to
+  use the new branch as the connection default.
+
+* It must have the ``--empty`` option. If specified, this is equivalent to
+  running ``create empty branch <newname>`` DDL command.
+
+* It should have the ``--copy-data`` option. By default a new branch copies
+  schema only.
+
+The old ``drop`` subcommand works pretty much same as before.
+
+The old ``wipe`` subcommand may still be relevant for resetting a particular
+branch.
+
+There must be a new ``set`` sub-command that allows changing the default
+branch to connect to by updating the ``edgedb.toml`` file.
+
+
+Implementation
+--------------
+
+Most of the ``branch`` functionality is either existing ``database``
+functionality or can be implemented on top of that by performing migrations.
+
+The option to copy the data can be implemented in Postgres by using the
+``CREATE DATABASE newname WITH TEMPLATE = oldname`` which may be preferable
+to a dump/restore as it should be faster. The caveat is that unlike with a
+dump/restore the template database cannot have any other active connections.
+However, this may be reasonable for local development as the benefit is speed.
+
+
+Rebasing Branches
+-----------------
+
+We need to be able to rebase and merge database branches. This overlaps a lot
+with the scope of the ``migration`` commands.
+
+When rebasing one branch on top of another we can use introspection to compare
+the respective migration histories and find the point where they diverge.
+Afterwards we can try to apply a batch of new migration to the existing branch
+and if there are no issues perform a "fast-forward" rebase.
+
+In order to minimize the hassle of rebasing the git branches corresponding to
+database branches we need to give migration files names that are distinct in
+these branches so that git does not attempt to merge the file contents. This
+may be either directly the branch name appended to the numeric name or a
+(short) hash derived from the branch name.
```

Comment:
We can also potentially avoid needing to have distinct migration file names by having a little bit of VCS integration. Not sure if we want to, though?

Reply:
I think VCS integration should be strictly optional.

> This may be either directly the branch name appended to the numeric name or a (short) hash derived from the branch name.

I think @elprans proposed hashing the message part of the migration?

Reply:
What's the problem with branch names in migration files?

Just by looking at the file system, it is hard to determine which migration is the "last" one and should match with the state of the database?

Reply:
The issue is that when we are doing an `edgedb migration rebase` (or whatever), it will probably be while trying to resolve a `git rebase`/`git merge`. Since both branches will have migrations not in the other in the interesting case, we will have `00010.edgeql` (or whatever) filled with merge conflict markers, and so we can't necessarily reconstruct anything from it.

That said, I don't totally remember why exactly we decided we needed information from the file system in order to do an `edgedb migration rebase`. We can read the migration histories out of the database branches themselves, possibly, and find the common ancestor that way?

I have a nagging feeling that we found some fatal flaw in that plan, though, but I can't remember it, and I think I left all of my notes from the offsite in my hotel room...